In [12]:
import numpy as np
import pandas as pd

# EDA
train = pd.read_csv("/kaggle/input/titanic/train.csv")
#train.head
train.info
# train.shape
# train.isnull().sum()
# train['Embarked'].value_counts()
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [13]:
train['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

549 passengers died, and 342 passengers survived. Of those survivng passengers, what characteristics did they have that may have contributed to their survival? 

In [14]:
train['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

64% of the passengers are male, and 35% of the passengers are female. Because there are more male passengers, it can be mistakenly concluded that being male came with a higher chance of surviving/not surviving. We wouldn't necessarily know if the same could be said for female passengers, as there isn't an equal amount of them and therefore their results to compare to males, and vice-versa.

In [15]:
(train
 .groupby(['Sex','Survived'])
 .agg(
     tot_surv=pd.NamedAgg('Survived', 'count')))

tot_surv
Sex    Survived          
female 0               81
       1              233
male   0              468
       1              109

In [16]:
train['Age'].describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

In [17]:
train[train['Age']<=28].count()

PassengerId    362
Survived       362
Pclass         362
Name           362
Sex            362
Age            362
SibSp          362
Parch          362
Ticket         362
Fare           362
Cabin           60
Embarked       362
dtype: int64

In [20]:
class Quantile:
    def __init__(self, q):
        self.q=q
    def __call__(self, x):
        return x.quantile(self.q)
    
print(
train
.groupby(['Sex', 'Survived'])
.agg(
    b1=('Age', Quantile(0.25)),
    b2=('Age', Quantile(0.5)),
    b3=('Age', Quantile(0.75))
    )
)


                    b1    b2     b3
Sex    Survived                    
female 0         16.75  24.5  33.25
       1         19.00  28.0  38.00
male   0         21.75  29.0  39.25
       1         18.00  28.0  36.00
